design/design-doc-webhooks.md

```diff
@@ -0,0 +1,383 @@
+    queries the value of that field needs to be known, hence we need to make a
+    `GET` call to the provider API, which means we'd have to import GCP Go SDK.
+    We update SDKs very infrequently so we have not encountered this but it's
+    possible.
+  * If we decide to expand abilities of webhooks, we might need other libraries.
+    For example, if admission webhook has a functionality checking the
+    `ServiceAccount` of the action taker, then we probably need to call
+    api-server to get its RBAC details, which would require importing `client`
+    package and some cumbersome logic to make it available in the CRD struct.
+* `Default` is not a well-suited name for our mutation operations we expect to
+  do since we'll use the kubebuilder [`default`
+  marker](https://book.kubebuilder.io/reference/markers/crd-validation.html) for
+  this functionality.
+
+On the other hand, while we could work with the low level one, we never really
+need to customize encoder/decoders and deal with `http.Handler` interface. We'd
+need to build a layer similar to the one we described here as high level
+abstraction but in `crossplane-runtime`.
+
+## Proposal
+
+There are four main pieces of a webhook:
+* Actual Go implementation of the logic.
+* Exposing an HTTP server accepting requests from Kubernetes API Server and
+  responding properly.
+* Handing off certificates to both API Server and the webhook server since TLS
+  is required.
+* Making sure the server is registered to Kubernetes API Server.
+
+Because of the limitations of the `controller-runtime` high-level abstraction,
+we will go with the low level abstraction and build the necessary structs in
+`crossplane-runtime` that will be used in providers.
+
+### The Abstraction Layer
+
+In `crossplane-runtime`, we will accept a list of functions for each type of
+the webhook and make sure they are called in given order whenever a request
+hits. The main difference between `controller-runtime` is that these functions
+do not have to be members of API structs; they can belong to any object or just
+be anonymous functions. Though the recommendation would be to have a struct
+similar to implementations of `ExternalClient` where all functions are provided
+by a single struct.
+
+### Exposing Webhook Server
+
+When a controller-runtime manager starts, the webhook server is also started
+automatically. So, as long as necessary implementations are there and hooked up
+with the main manager object, we have the server up and running. But in order to
+expose it to cluster, we need to create a `Service` resource similar to
+`Deployment` of the controller.
+
+The package manager will create an opinionated `Service` resource for the
+providers it installs and this will require Crossplane to have necessary RBAC
+for managing `Service` objects, too. Though it could be limited to which
+namespace it's installed, again similar to `Deployment` RBAC.
```

Comment:
This seems like it could me a prime candidate for sitting behind the [Provider Runtime Interface](https://github.com/crossplane/crossplane/issues/2671) so that we are not explicitly tied to using a `Service` since [deploying outside a cluster](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#deploy-the-admission-webhook-service) is an option.

Reply:
If we do end up with that runtime interface, I agree that it could be a good candidate but right now, I'm planning to integrate to the existing mechanisms with as little change as possible.